pkg/boot/universalpayload/utilities.go

```diff
@@ -0,0 +1,139 @@
+
+type FdtLoad struct {
+	Load       uint64
+	EntryStart uint64
+}
+
+// GetFdtInfo Device Tree Blob resides at the start of FIT binary. In order to
+// get the expected load and entry point address, need to walk though
+// DTB to get value of properties 'load' and 'entry-start'.
+//
+// The simplified device tree layout is:
+//
+//	/{
+//	    images {
+//	        tianocore {
+//	            entry-start = <0x00000000 0x00805ac3>;
+//	            load = <0x00000000 0x00800000>;
+//	        }
+//	    }
+//	 }
+func GetFdtInfo(name string) (*FdtLoad, error) {
+	return getFdtInfo(name, nil)
+}
+
+func getFdtInfo(name string, dtb io.ReaderAt) (*FdtLoad, error) {
+	var fdt *dt.FDT
+	var err error
+
+	if dtb != nil {
+		fdt, err = dt.ReadFDT(io.NewSectionReader(dtb, 0, math.MaxInt64))
+	} else {
+		fdt, err = dt.ReadFile(name)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to read fdt file: %s, err = %w", name, err)
+	}
+
+	firstLevelNode, succeed := fdt.NodeByName(FirstLevelNodeName)
+	if succeed != true {
+		return nil, fmt.Errorf("failed to find '%s' node", FirstLevelNodeName)
+	}
+
+	secondLevelNode, succeed := firstLevelNode.NodeByName(SecondLevelNodeName)
+	if succeed != true {
+		return nil, fmt.Errorf("failed to find '%s' node", SecondLevelNodeName)
+	}
+
+	loadAddrProp, succeed := secondLevelNode.LookProperty(LoadAddrPropertyName)
+	if succeed != true {
+		return nil, fmt.Errorf("failed to find get '%s' property", LoadAddrPropertyName)
+	}
+
+	loadAddr, err := loadAddrProp.AsU64()
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert property '%s' to u64", LoadAddrPropertyName)
+	}
+
+	entryAddrProp, succeed := secondLevelNode.LookProperty(EntryAddrPropertyName)
+	if succeed != true {
+		return nil, fmt.Errorf("failed to find get '%s' property", EntryAddrPropertyName)
+	}
+
+	entryAddr, err := entryAddrProp.AsU64()
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert property '%s' to u64", EntryAddrPropertyName)
+	}
+
+	return &FdtLoad{
+		Load:       loadAddr,
+		EntryStart: entryAddr,
+	}, nil
+}
+
+// Get Physical Address size from sysfs node /proc/cpuinfo.
+// Both Physical and Virtual Address size will be prompted as format:
+// "address sizes	: 39 bits physical, 48 bits virtual"
+// Use regular expression to fetch the integer of Physical Address
+// size before "bits physical" keyword
+func getPhysicalAddressSizes() (uint8, error) {
+	file, err := os.Open(sysfsCPUInfoPath)
+	if err != nil {
+		return 0, fmt.Errorf("failed to open %s: %v", sysfsCPUInfoPath, err)
+	}
+	defer file.Close()
+
+	// Regular expression to match the address size line
+	re := regexp.MustCompile(`address sizes\s*:\s*(\d+)\s+bits physical,\s*(\d+)\s+bits virtual`)
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if match := re.FindStringSubmatch(line); match != nil {
+			// Convert the physical bits size to integer
+			physicalBits, err := strconv.Atoi(match[1])
```

Comment:
if you use StroToULL you can specify the allowed number of bits, and then you won't need lines 127-129